### DISCUSSION TITLE
Avoid update state of SN provider when a copy of the state (or part of it) is modified into another provider

### DISCUSSION THREAD

**Discussion by garcamcoder** [rrousselGit/riverpod#3506] [Q&A] [answered]
Hi, I am using an StateNotifier to handle a multi-screen form (for editing/updating records) in my app, and I'm having an issue that is making me break my mind.
(please asume that all imports are correctly added)

First of all my model called '**MyModel**' is something like this:

```
class MyModel extends Equatable {
  final String property1;
  final Map<String, dynamic> otherProperties;

  const MyModel({
    required this.property1,
    required this.otherProperties,
  });

  MyModel copyWith({
    String? property1,
    Map<String, dynamic>? otherProperties,
  }) {
    return MyModel(
      property1: property1 ?? this.property1,
      otherProperties: otherProperties ?? this.otherProperties,
    );
  }

  @override
  List<Object?> get props => throw UnimplementedError();
}
```

This model is used in an StateNotifier provider called '**myProvider**' that keeps avilable the registers across the app for displaying purpouses.

```
final myProvider = StateNotifierProvider<MyNotifier, MyState>(
  ((ref) => MyNotifier()),
);

class MyState extends Equatable {
  final String anotherProp;
  final List<MyModel> registers;

  const MyState._({
    required this.anotherProp,
    required this.registers,
  });

  const MyState.initial()
      : this._(
          anotherProp: '',
          registers: const [],
        );

  MyState copyWith({List<MyModel>? registers, String? anotherProp}) {
    return MyState._(
      anotherProp: anotherProp ?? this.anotherProp,
      registers: registers ?? this.registers,
    );
  }

  @override
  List<Object?> get props => [anotherProp, registers];
}

class MyNotifier extends StateNotifier<MyState> {
  MyNotifier() : super(const MyState.initial());

  void updateRegister(MyModel myModelData, int index) {
    final auxRegisters = state.registers;

    auxRegisters[index] = myModelData;

    state = MyState._(
      anotherProp: state.anotherProp,
      registers: auxRegisters,
    );
  }
}
```

Multi-screen edition form is being handle by an StateNotifier like this called '**editionProvider**'

```
final editionProvider = StateNotifierProvider<EditionProvider, EditionState>(
  ((ref) => EditionProvider(
        ref.watch(myProvider),
      )),
);

class EditionState extends Equatable {
  final String edProperty1;
  final Map<String, dynamic> edOtherProperties;

  const EditionState._({
    required this.edProperty1,
    required this.edOtherProperties,
  });

  const EditionState.empty()
      : this._(
          edProperty1: '',
          edOtherProperties: const {},
        );

  EditionState copyWith({
    String? property1,
    Map<String, dynamic>? otherProperties,
  }) {
    return EditionState._(
      edProperty1: property1 ?? this.edProperty1,
      edOtherProperties: otherProperties ?? this.edOtherProperties,
    );
  }

  @override
  List<Object?> get props => [edProperty1, edOtherProperties];
}

class EditionProvider extends StateNotifier<EditionState> {
  final MyState _myState;

  EditionProvider(
    this._myState,
  ) : super(const EditionState.empty());

  void createRegisterFromExisting(int index) {
    final originalRegister = _myState.registers[index];

    state = EditionState._(
      edProperty1: originalRegister.property1,
      edOtherProperties: originalRegister.otherProperties,
    );
  }

  void modifyPropertiesList(String propertyKey, dynamic propertyValue) {
    Map<String, dynamic> updatedPropertiesList = state.edOtherProperties;

    updatedPropertiesList[propertyKey] = propertyValue;

    state = state.copyWith(
      otherProperties: updatedPropertiesList,
    );
  }

  void saveNewRegister() {
    //Some code to add new register instance to the list stored into
    //myProvider's state using the data handled in this provider's state
  }
}
```

My problem comes in the methods '**createRegisterFromeExisting()**' and '**modifyPropertiesList()**', in the firstone I make a copy of a map called '**otherProperties**' from a MyModel instance from the list stored in **myProvider**'s state to a local property of **editionProvider**'s state called '**edOtherProperties**'. So when I call modifyPropertiesList method to update the local copy of the map (**edOtherProperties**) the change is reflected also into the original source of the map stored in the instance of '**otherProperties**' of the liste element stored in **myProvider**'s state.

So, I deduce that '**otherProperties**' (from the list element from in **myProvider**'s state) is being referenced from '**edOtherProperties**' (from **editionProvider**'s state), but I dont know how to break this reference or avoid the undesierd modification of the original map source.

Thanks in advance.

**Comment by gdurandrexel** [comment]
In `createRegisterFromeExisting`, `edOtherProperties: originalRegister.otherProperties` does not create a copy of otherProperties. It just assigns the reference of otherProperties to edOtherProperties, so both originalRegister.otherProperties and edOtherProperties reference the same object.
If you want to make a copy, use `edOtherProperties: Map.from(originalRegister.otherProperties)`. But note that this is a shallow copy: the keys and values will reference the same objects. If you just want to add/remove items in the map that's fine, but if you update the content of some values, then it will be "reflected" in the original model.

**Reply by garcamcoder** [reply, marked as answer]
Thank you very much, after your explanation (it was very illustrative) I remembered that I defined 2 methods in all my models (wich are structured as DataClasses extended from Equatable), **.toMap()** to export the instance as a map, and also **.fromMap()** to create a new instance from a map. So, the way I found to make sure I created an unreferenced copy of my original variable was to use these 2 methods, in my example it would be something like:

```
void createRegisterFromExisting(int index) {

    final copyOfOriginalRegister = MyModel.fromMap(_myState.registers[index].toMap());

    state = EditionState._(
      edProperty1: copyOfOriginalRegister.property1,
      edOtherProperties: copyOfOriginalRegister.otherProperties,
    );
  }
```
